src/Components/Endpoints/src/Binding/DefaultFormValuesSupplier.cs

```diff
@@ -14,69 +14,96 @@ namespace Microsoft.AspNetCore.Components.Endpoints;
+        var deserializer = _cache.GetOrAdd(context.ValueType, CreateDeserializer);
+        Debug.Assert(deserializer != null);
 
-        var result = deserializer(_formData.Entries, _options, "value");
-        if (result != default)
-        {
-            // This is not correct, but works for primtive values.
-            // Will change the interface when we add support for complex types.
-            boundValue = result;
-            return true;
-        }
-
-        boundValue = valueType.IsValueType ? Activator.CreateInstance(valueType) : null;
-        return false;
+        deserializer.Deserialize(context, _options, _formData.Entries);
     }
 
-    private Func<IReadOnlyDictionary<string, StringValues>, FormDataMapperOptions, string, object> CreateDeserializer(Type type) =>
-        _method.MakeGenericMethod(type)
-        .CreateDelegate<Func<IReadOnlyDictionary<string, StringValues>, FormDataMapperOptions, string, object>>();
+    private FormValueSupplier CreateDeserializer(Type type) =>
+        (FormValueSupplier)Activator.CreateInstance(typeof(FormValueSupplier<>)
+        .MakeGenericType(type))!;
 
-    private static object? DeserializeCore<T>(IReadOnlyDictionary<string, StringValues> form, FormDataMapperOptions options, string value)
+    internal abstract class FormValueSupplier
     {
-        // Form values are parsed according to the culture of the request, which is set to the current culture by the localization middleware.
-        // Some form input types use the invariant culture when sending the data to the server. For those cases, we'll
-        // provide a way to override the culture to use to parse that value.
-        var reader = new FormDataReader(form, CultureInfo.CurrentCulture);
-        reader.PushPrefix(value);
-        return FormDataMapper.Map<T>(reader, options);
+        public abstract void Deserialize(FormValueSupplierContext context, FormDataMapperOptions options, IReadOnlyDictionary<string, StringValues> form);
     }
 
-    public bool CanConvertSingleValue(Type type)
+    internal class FormValueSupplier<T> : FormValueSupplier
     {
-        return _options.IsSingleValueConverter(type);
+        public override void Deserialize(FormValueSupplierContext context, FormDataMapperOptions options, IReadOnlyDictionary<string, StringValues> form)
+        {
+            if (form.Count == 0)
+            {
+                return;
+            }
+
+            char[]? buffer = null;
+            try
+            {
+                var maxKeyLenght = -1;
+                var dictionary = new Dictionary<FormKey, StringValues>();
+                foreach (var (key, value) in form)
+                {
+                    if (key.Length > maxKeyLenght)
+                    {
+                        maxKeyLenght = key.Length;
+                    }
+                    dictionary.Add(new FormKey(key.AsMemory()), value);
+                }
+                buffer = ArrayPool<char>.Shared.Rent(maxKeyLenght);
```

Comment:
Is it worth pooling here? It looks like we're only taking an array whose size is <= the biggest key we're already holding, so it can't be a huge array or at least no bigger than data we already have.

Not a big deal, just wondering if we can avoid having to concern ourselves with the impact on contention across requests.

Reply:
I don't think we should worry ourselves about contention here.

I've changed the logic here actually to use the limits from the form options, as that's more correct.